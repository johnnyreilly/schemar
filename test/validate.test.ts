import fs from "node:fs/promises";
import { describe, expect, it } from "vitest";
import {
	processValidationResponse,
	processValidationResult,
} from "../src/validate.js";
import path from "node:path";

const prefix = `)]}'
`;

describe("processValidationResult", () => {
	it("given valid schema without errors / warnings returns successfully", async () => {
		const text =
			prefix +
			(await fs.readFile(
				path.join(__dirname, "exampleValidStructuredDataValidation.json"),
				"utf8",
			));

		const validationResult = processValidationResponse(
			"https://deploy-preview-9669--docusaurus-2.netlify.app/blog/releases/2.4/",
			text,
		);
		const processedValidationResult = processValidationResult(validationResult);

		expect(processedValidationResult.success).toBe(true);
		expect(processedValidationResult).toMatchSnapshot();
	});

	it("given errors returns unsuccessfully", async () => {
		const text =
			prefix +
			(await fs.readFile(
				path.join(__dirname, "exampleInvalidStructuredDataValidation.json"),
				"utf8",
			));

		const validationResult = processValidationResponse(
			"https://thankful-sky-0bfc7e803-804.westeurope.1.azurestaticapps.net/",
			text,
		);
		if (typeof validationResult === "string") {
			throw new Error("validationResult is string");
		}
		validationResult.totalNumErrors = 1;
		const processedValidationResult = processValidationResult(validationResult);

		expect(processedValidationResult.success).toBe(false);
		expect(processedValidationResult).toMatchSnapshot();
	});

	it("given warnings returns unsuccessfully", async () => {
		const text =
			prefix +
			(await fs.readFile(
				path.join(__dirname, "exampleValidStructuredDataValidation.json"),
				"utf8",
			));

		const validationResult = processValidationResponse(
			"https://deploy-preview-9669--docusaurus-2.netlify.app/blog/releases/2.4/",
			text,
		);
		if (typeof validationResult === "string") {
			throw new Error("validationResult is string");
		}
		validationResult.totalNumWarnings = 1;
		const processedValidationResult = processValidationResult(validationResult);

		expect(processedValidationResult.success).toBe(false);
		expect(processedValidationResult).toMatchSnapshot();
	});

	it("given no structured data returns unsuccessfully", async () => {
		const text =
			prefix +
			(await fs.readFile(
				path.join(__dirname, "exampleNoStructuredDataValidation.json"),
				"utf8",
			));

		const validationResult = processValidationResponse(
			"https://news.ycombinator.com/",
			text,
		);
		const processedValidationResult = processValidationResult(validationResult);

		expect(processedValidationResult.success).toBe(false);
		expect(processedValidationResult).toMatchSnapshot();
	});
});
