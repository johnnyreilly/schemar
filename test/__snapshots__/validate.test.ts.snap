// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`processValidationResult > given errors returns unsuccessfully 1`] = `
{
  "resultText": "Validated https://thankful-sky-0bfc7e803-804.westeurope.1.azurestaticapps.net/ and failed with 1 errors and 0 warnings
For more details see https://validator.schema.org/#url=https%3A%2F%2Fthankful-sky-0bfc7e803-804.westeurope.1.azurestaticapps.net%2F
",
  "success": false,
}
`;

exports[`processValidationResult > given no structured data returns unsuccessfully 1`] = `
{
  "resultText": "Validated https://news.ycombinator.com/ and found no structured data
For more details see https://validator.schema.org/#url=https%3A%2F%2Fnews.ycombinator.com%2F
",
  "success": false,
}
`;

exports[`processValidationResult > given valid schema without errors / warnings returns successfully 1`] = `
{
  "resultText": "https://deploy-preview-9669--docusaurus-2.netlify.app/blog/releases/2.4/ has structured data of these types:
 - BlogPosting

For more details see https://validator.schema.org/#url=https%3A%2F%2Fdeploy-preview-9669--docusaurus-2.netlify.app%2Fblog%2Freleases%2F2.4%2F
",
  "success": true,
}
`;

exports[`processValidationResult > given warnings returns unsuccessfully 1`] = `
{
  "resultText": "Validated https://deploy-preview-9669--docusaurus-2.netlify.app/blog/releases/2.4/ and failed with 0 errors and 1 warnings
For more details see https://validator.schema.org/#url=https%3A%2F%2Fdeploy-preview-9669--docusaurus-2.netlify.app%2Fblog%2Freleases%2F2.4%2F
",
  "success": false,
}
`;
